File: c++/c08_2_nearest_smaller_elements.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	vector<ll> solve(const vector<ll>& a) {
		stack<ll> s;
		vector<ll> ret;
		for (ll i = 0; i < ssize(a); ++i) {
			while (!s.empty() && a[s.top()] >= a[i])
				s.pop();
			ret.push_back((!s.empty()) ? s.top() : -1);
			s.push(i);
		}
		return ret;
	}
}

TEST(C082NearestSmallerElements, solve) {
	EXPECT_EQ(solve({}), vector<ll>{});
	EXPECT_EQ(solve({ 1,3,4,2,5,3,4,2 }), (vector<ll>{ -1,0,1,0,3,3,5,0 }));
}

File: python/c08_2_nearest_smaller_elements.py
import unittest

def solve(a):
    s=[]
    for i in range(len(a)):
        while s and a[s[-1]]>=a[i]:
            s.pop()
        yield s[-1] if s else None
        s.append(i)

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual([*solve([])],[])
        self.assertEqual([*solve([1,3,4,2,5,3,4,2])],[None,0,1,0,3,3,5,0])

if __name__=='__main__':
    unittest.main()
